display_list/geometry/dl_region.h

```diff
@@ -28,36 +38,111 @@ class DlRegion {
   /// merged into single rectange.
   std::vector<SkIRect> getRects(bool deband = true) const;
 
+  /// Returns maximum and minimum axis values of rectangles in this region.
+  /// If region is empty returns SKIRect::MakeEmpty().
+  const SkIRect& bounds() const { return bounds_; }
+
+  /// Returns whether this region intersects with a rectangle.
+  bool intersects(const SkIRect& rect) const;
+
+  /// Returns whether this region intersects with another region.
+  bool intersects(const DlRegion& region) const;
+
+  ~DlRegion();
+
  private:
-  void addRects(std::vector<SkIRect>&& rects);
+  typedef std::uint32_t SpanChunkHandle;
 
   struct Span {
     int32_t left;
     int32_t right;
+
+    Span() = default;
+    Span(int32_t left, int32_t right) : left(left), right(right) {}
   };
-  typedef std::vector<Span> SpanVec;
+
+  /// Holds spans for the region. Having custom allocated memory that doesn't
+  /// do zero initialization every time the buffer gets resized improves
+  /// performance measurably.
+  class SpanBuffer {
+   public:
+    SpanBuffer() = default;
+    SpanBuffer(const SpanBuffer&) = delete;
+    SpanBuffer(SpanBuffer&& m);
+
+    void reserve(size_t capacity);
+    size_t capacity() const { return capacity_; }
+
+    SpanChunkHandle storeChunk(const Span* begin, const Span* end);
+    size_t getChunkSize(SpanChunkHandle handle) const;
+    void getSpans(SpanChunkHandle handle,
+                  const DlRegion::Span*& begin,
+                  const DlRegion::Span*& end) const;
+
+    ~SpanBuffer();
+
+   private:
+    size_t capacity_ = 0;
+    size_t size_ = 0;
+
+    // Spans for the region chunks. First span in each chunk contains the
+    // chunk size.
+    Span* spans_ = nullptr;
+  };
+
+  DlRegion();
+
+  bool isEmpty() const { return lines_.empty(); }
+  bool isComplex() const;
+
   struct SpanLine {
     int32_t top;
     int32_t bottom;
-    SpanVec* spans;
-
-    void insertSpan(int32_t left, int32_t right);
-    bool spansEqual(const SpanLine& l2) const;
+    SpanChunkHandle chunk_handle;
   };
 
-  typedef std::vector<SpanLine> LineVec;
+  void addRects(const std::vector<SkIRect>& rects);
 
-  std::vector<SpanLine> lines_;
-  std::vector<SpanVec*> spanvec_pool_;
-
-  void insertLine(size_t position, SpanLine line);
-  LineVec::iterator removeLine(LineVec::iterator position);
+  void appendLine(int32_t top,
+                  int32_t bottom,
+                  const Span* begin,
+                  const Span* end);
+  void appendLine(int32_t top,
+                  int32_t bottom,
+                  const SpanBuffer& buffer,
+                  SpanChunkHandle handle) {
+    const Span *begin, *end;
+    buffer.getSpans(handle, begin, end);
+    appendLine(top, bottom, begin, end);
+  }
 
+  typedef std::vector<Span> SpanVec;
```

Comment:
(const) vector of `const Span`?

Reply:
I don't think `emplace_back` will compile with `std::vector<const Span>`:

```cxx
std::vector<const Span> vec;
vec.emplace_back(1, 2);
```